blockprod/src/detail/timestamp_searcher.rs

```diff
@@ -0,0 +1,487 @@
+                            let prev_block_index =
+                                get_existing_gen_block_index(chainstate, &prev_block_id)?;
+
+                            prev_block_index.block_timestamp()
+                        };
+
+                    let seconds_to_check =
+                        if cur_height == max_height || !check_all_timestamps_between_blocks {
+                            seconds_to_check_for_height
+                        } else {
+                            assert!(block_index.block_timestamp() > prev_block_timestamp);
+                            block_index.block_timestamp().as_int_seconds()
+                                - prev_block_timestamp.as_int_seconds()
+                                - 1
+                        };
+
+                    block_index_from_last_iter = Some(block_index);
+
+                    (target_required, prev_block_timestamp, seconds_to_check)
+                };
+
+            let min_timestamp = timestamp_add_secs(prev_block_timestamp, 1)?;
+            let max_timestamp = timestamp_add_secs(min_timestamp, seconds_to_check)?;
+            let consensus_version = pos_status.get_chain_config().consensus_version();
+
+            let data_for_height = SearchDataForHeight {
+                sealed_epoch_randomness,
+                epoch_index,
+                target_required,
+                min_timestamp,
+                max_timestamp,
+                pool_balances: cur_pool_balances,
+                consensus_version,
+            };
+
+            search_data.push(data_for_height);
+        }
+
+        Ok(Self {
+            start_height: min_height,
+            data: search_data,
+            check_all_timestamps_between_blocks,
+        })
+    }
+}
+
+/// Collect search data that can be subsequently passed into find_timestamps_for_staking
+/// to perform timestamps search.
+///
+/// For each block height in the specified range, the search will find timestamps where staking
+/// is/was possible.
+///
+/// `min_height` must not be zero; `max_height` must not exceed the best block height plus one
+/// (this value is assumed if `None` is passed).
+///
+/// If `check_all_timestamps_between_blocks` is false, `seconds_to_check_for_height + 1` is
+/// the number of seconds that will be checked at each height in the range.
+/// If `check_all_timestamps_between_blocks` is true, `seconds_to_check_for_height` only applies
+/// to the last height in the range; for all other heights the maximum timestamp is the timestamp
+/// of the next block.
+pub async fn collect_timestamp_search_data(
+    chainstate_handle: &ChainstateHandle,
+    chain_config: Arc<ChainConfig>,
+    secret_input_data: PoSTimestampSearchInputData,
+    min_height: BlockHeight,
+    max_height: Option<BlockHeight>,
+    seconds_to_check_for_height: u64,
+    check_all_timestamps_between_blocks: bool,
+) -> Result<TimestampSearchData, BlockProductionError> {
```

Comment:
2discuss: the `check_all_timestamps_between_blocks` parameter seems redundant here, probably it should be removed and the code should behave as if it was always true?

Reply:
This is an auxiliary functionality, so I'm OK with it not being perfect. It's only gonna be used by us, devs, mostly.